internal/oci/build.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oci
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// Build archives the given directory as a tarball to the given local path.
+// While archiving, any environment specific data (for example, the user and group name) is stripped from file headers.
+func Build(artifactPath, sourceDir string) error {
+	if f, err := os.Stat(sourceDir); os.IsNotExist(err) || !f.IsDir() {
+		return fmt.Errorf("invalid source dir path: %s", sourceDir)
+	}
+
+	tf, err := os.CreateTemp(filepath.Split(sourceDir))
+	if err != nil {
+		return err
+	}
+	tmpName := tf.Name()
+	defer func() {
+		if err != nil {
```

Comment:
this evaluate the error when the defer is declared, not sure that's what intended. Maybe a better option is to have a named return parameter

Reply:
This is taken from there https://github.com/fluxcd/source-controller/blob/main/controllers/storage.go#L345-L349 why is it better to have a named return parameter?

Reply:
Ah I get what you're saying, will fix ty.